price-feeder/oracle/provider/huobi.go

```diff
@@ -27,140 +32,248 @@ type (
+		reconnectTicker: time.NewTicker(huobiReconnectTime),
+	}
+
+	if err := provider.subscribeTickers(pairs...); err != nil {
+		return nil, err
+	}
+	provider.subscribedPairs = pairs
+
+	go provider.handleWebSocketMsgs(ctx)
+
+	return provider, nil
 }
 
-func (p HuobiProvider) GetTickerPrices(pairs ...types.CurrencyPair) (map[string]TickerPrice, error) {
-	tickerPrices := make(map[string]TickerPrice, len(pairs))
-	for _, cp := range pairs {
-		price, err := p.getTickerPrice(cp.String())
-		if err != nil {
-			return nil, err
+// subscribeTickers subscribe to all currency pairs.
+func (p *HuobiProvider) subscribeTickers(cps ...types.CurrencyPair) error {
+	for _, cp := range cps {
+		huobiSubscriptionMsg := newHuobiSubscriptionMsg(cp)
+		if err := p.wsClient.WriteJSON(huobiSubscriptionMsg); err != nil {
+			return err
 		}
-
-		tickerPrices[cp.String()] = price
 	}
 
-	return tickerPrices, nil
+	return nil
 }
 
-func (p HuobiProvider) getTickerPrice(ticker string) (TickerPrice, error) {
-	path := fmt.Sprintf("%s%s?symbol=%s", p.baseURL, huobiTradeEndpoint, strings.ToLower(ticker))
+func (p *HuobiProvider) handleWebSocketMsgs(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-time.After(defaultReadNewMessage):
+			// time after to avoid asking for prices too frequently
+			messageType, bz, err := p.wsClient.ReadMessage()
+			if err != nil {
+				// if some error occurs continue to try to read the next message
+				p.logger.Err(err).Msg("Huobi provider could not read message")
+				if err := p.ping(); err != nil {
+					p.logger.Err(err).Msg("Error sending ping")
+					if err := p.reconnect(); err != nil {
+						p.logger.Err(err).Msg("error reconnecting to the huobi provider")
+					}
+				}
+				continue
+			}
 
-	resp, err := p.client.Get(path)
-	if err != nil {
-		return TickerPrice{}, fmt.Errorf("failed to make Huobi request: %w", err)
+			if len(bz) == 0 {
+				continue
+			}
+
+			p.messageReceived(messageType, bz)
+
+		case <-p.reconnectTicker.C:
+			if err := p.reconnect(); err != nil {
+				p.logger.Err(err).Msg("error reconnecting to the huobi provider")
+			}
+		}
 	}
+}
 
-	defer resp.Body.Close()
+// messageReceived handles the received data from Huobi
+// All return data of websocket Market APIs are compressed with
+// GZIP so they need to be unzipped.
+func (p *HuobiProvider) messageReceived(messageType int, bz []byte) {
+	if messageType != websocket.BinaryMessage {
+		return
+	}
 
-	bz, err := ioutil.ReadAll(resp.Body)
+	bz, err := uncompressGzip(bz)
 	if err != nil {
-		return TickerPrice{}, fmt.Errorf("failed to read Huobi response body: %w", err)
+		p.logger.Err(err).Msg("Huobi provider could not uncompress gzip message")
+		return
 	}
 
-	var tradeResp HuobiTraceResponse
-	if err := json.Unmarshal(bz, &tradeResp); err != nil {
-		return TickerPrice{}, fmt.Errorf("failed to unmarshal Huobi response body: %w", err)
+	if bytes.Contains(bz, []byte("ping")) {
```

Comment:
let's put `[]byte("ping")` in a global private `var`?

Reply:
done at https://github.com/umee-network/umee/pull/569/commits/e9099221b3884748aed6f8862b70836b3a2270e7